components/gfx/display_list/mod.rs

```diff
@@ -646,6 +666,81 @@ impl StackingContext {
     }
 }
 
+struct StackingContextLayerizer {
+    display_list_for_next_layer: Option<Box<DisplayList>>,
+    all_following_children_need_layers: bool,
+}
+
+impl StackingContextLayerizer {
+    fn new() -> StackingContextLayerizer {
+        StackingContextLayerizer {
+            display_list_for_next_layer: None,
+            all_following_children_need_layers: false,
+        }
+    }
+
+    #[inline]
+    fn layerize(stacking_context: &mut StackingContext) {
+        let mut state = StackingContextLayerizer::new();
+        let mut children: SmallVec<[Arc<StackingContext>; 8]> = SmallVec::new();
+        while let Some(child_stacking_context) = stacking_context.display_list.children.pop_front() {
```

Comment:
This could be a (long) one-liner if you did something like `children.extend(mem::replace(&mut stacking_context.display_list.children, LinkedList::new()).into_iter())` (assuming `SmallVec` has `extend`), but this is fairly clear as-is, so it's up to you.